File: src/engine/player.ts
// src/engine/player.ts
//
// Direct port of your Java Player, minus console printing.
// UI should render hand/groups.

import { Card } from "./card";

export class Player {
  private hand: Card[] = [];
  private opened = false;
  private openGroups: Card[][] = [];

  public addCard(c: Card): void {
    this.hand.push(c);
  }

  public removeCard(c: Card): void {
    const idx = this.hand.indexOf(c);
    if (idx >= 0) this.hand.splice(idx, 1);
  }

  public getHand(): Card[] {
    // return a copy to avoid UI mutating engine state accidentally
    return [...this.hand];
  }

  public hasOpened(): boolean {
    return this.opened;
  }

  public open(): void {
    this.opened = true;
  }

  public addGroup(g: ReadonlyArray<Card>): void {
    this.openGroups.push([...g]);
  }

  public getGroups(): Card[][] {
    // shallow copy of groups
    return this.openGroups.map((g) => [...g]);
  }
}

File: Player.java
package card51;

import java.util.*;

public class Player {
    private final List<Card> hand = new ArrayList<>();
    private boolean opened = false;
    private final List<List<Card>> openGroups = new ArrayList<>();

    public void addCard(Card c) {
        hand.add(c);
    }

    public void removeCard(Card c) {
        hand.remove(c);
    }

    public List<Card> getHand() {
        return hand;
    }

    public boolean hasOpened() {
        return opened;
    }

    public void open() {
        opened = true;
    }

    public void addGroup(List<Card> g) {
        openGroups.add(new ArrayList<>(g));
    }

    public List<List<Card>> getGroups() {
        return openGroups;
    }

    public void showHand() {
        System.out.println("Your hand:");
        for (int i = 0; i < hand.size(); i++) {
            System.out.printf("%2d: %s%n", i + 1, hand.get(i));
        }
    }
}